subprojects/GOPATH/src/github.com/microsoft/hivedscheduler/pkg/api/types.go

```diff
@@ -158,3 +158,104 @@ type LazyPreemptionStatus struct {
 	// It was lazy preempted at PreemptionTime.
 	PreemptionTime meta.Time `json:"preemptionTime"`
 }
+
+type CellState string
+
+const (
+	CellFree CellState = "Free"
+	CellUsed CellState = "Used"
+)
+
+type CellHealthiness string
+
+const (
+	CellHealthy CellHealthiness = "Healthy"
+	CellBad     CellHealthiness = "Bad"
+)
+
+type CellStatus struct {
+	GpuType  string   `json:"gpuType,omitempty"`
+	CellType CellType `json:"cellType"`
+	// Address of a physical cell consists of its address (or index) in each level
+	// (e.g., node0/0/0/0 may represent node0, CPU socket 0, PCIe switch 0, GPU 0.
+	// Address of a virtual cell consists of its VC name, index of the preassigned cell,
+	// and the relative index in each level inside the preassigned cell
+	// (e.g., VC1/0/0 may represent VC1, preassigned cell 0, index 0 among its children)
+	CellAddress     CellAddress     `json:"cellAddress"`
+	CellState       CellState       `json:"cellState"`
+	CellHealthiness CellHealthiness `json:"cellHealthiness"`
+	CellPriority    int32           `json:"cellPriority"`
+}
+
+type PhysicalCellStatus struct {
+	CellStatus
+	CellChildren []*PhysicalCellStatus `json:"cellChildren,omitempty"`
+	VC           VirtualClusterName    `json:"vc,omitempty"`
+	VirtualCell  *VirtualCellStatus    `json:"virtualCell,omitempty"`
+}
+
+func (pcs *PhysicalCellStatus) copy() *PhysicalCellStatus {
+	copied := &PhysicalCellStatus{
```

Comment:
Does golang has deepcopy func for any object?
Check built-in or 3rd party lib? So you no need to impl by yourself, and hard to maintain

Reply:
No built-in. As far as I know programmers usually implement deep copy theirselves when the struct is simple. Otherwise they use 3rd party libs.
I think this struct is simple enough. I can switch to a 3rd party lib if you think is better.